RecoEgamma/EgammaElectronProducers/plugins/GsfElectronCoreEcalDrivenProducer.cc

```diff
@@ -1,79 +1,102 @@
-    : GsfElectronCoreBaseProducer(config) {}
+    : useGsfPfRecTracks_(config.getParameter<bool>("useGsfPfRecTracks")),
+      gsfPfRecTracksToken_(mayConsume<GsfPFRecTrackCollection>(config.getParameter<edm::InputTag>("gsfPfRecTracks"))),
+      gsfTracksToken_(consumes<reco::GsfTrackCollection>(config.getParameter<edm::InputTag>("gsfTracks"))),
+      ctfTracksToken_(consumes<reco::TrackCollection>(config.getParameter<edm::InputTag>("ctfTracks"))),
+      putToken_(produces<reco::GsfElectronCoreCollection>()) {}
 
-void GsfElectronCoreEcalDrivenProducer::produce(edm::Event& event, const edm::EventSetup& setup) {
-  // base input
-  GsfElectronCoreBaseProducer::initEvent(event, setup);
+void GsfElectronCoreEcalDrivenProducer::produce(edm::StreamID, edm::Event& event, const edm::EventSetup& setup) const {
+  auto gsfTracksHandle = event.getHandle(gsfTracksToken_);
+  auto ctfTracksHandle = event.getHandle(ctfTracksToken_);
 
   // output
-  auto electrons = std::make_unique<GsfElectronCoreCollection>();
+  reco::GsfElectronCoreCollection electrons;
 
   // loop on ecal driven tracks
   if (useGsfPfRecTracks_) {
-    const GsfPFRecTrackCollection* gsfPfRecTrackCollection = gsfPfRecTracksH_.product();
-    GsfPFRecTrackCollection::const_iterator gsfPfRecTrack;
-    for (gsfPfRecTrack = gsfPfRecTrackCollection->begin(); gsfPfRecTrack != gsfPfRecTrackCollection->end();
-         ++gsfPfRecTrack) {
-      const GsfTrackRef gsfTrackRef = gsfPfRecTrack->gsfTrackRef();
-      produceEcalDrivenCore(gsfTrackRef, electrons.get());
+    for (auto const& gsfPfRecTrack : event.get(gsfPfRecTracksToken_)) {
+      produceEcalDrivenCore(gsfPfRecTrack.gsfTrackRef(), electrons, ctfTracksHandle);
     }
   } else {
-    const GsfTrackCollection* gsfTrackCollection = gsfTracksH_.product();
-    for (unsigned int i = 0; i < gsfTrackCollection->size(); ++i) {
-      const GsfTrackRef gsfTrackRef = edm::Ref<GsfTrackCollection>(gsfTracksH_, i);
-      produceEcalDrivenCore(gsfTrackRef, electrons.get());
+    for (unsigned int i = 0; i < gsfTracksHandle->size(); ++i) {
+      produceEcalDrivenCore(edm::Ref<GsfTrackCollection>(gsfTracksHandle, i), electrons, ctfTracksHandle);
     }
   }
 
-  event.put(std::move(electrons));
+  event.emplace(putToken_, std::move(electrons));
 }
 
-void GsfElectronCoreEcalDrivenProducer::produceEcalDrivenCore(const GsfTrackRef& gsfTrackRef,
-                                                              GsfElectronCoreCollection* electrons) {
-  GsfElectronCore* eleCore = new GsfElectronCore(gsfTrackRef);
+void GsfElectronCoreEcalDrivenProducer::produceEcalDrivenCore(
+    const reco::GsfTrackRef& gsfTrackRef,
+    reco::GsfElectronCoreCollection& electrons,
+    edm::Handle<TrackCollection> const& ctfTracksHandle) const {
+  electrons.emplace_back(gsfTrackRef);
+  auto& eleCore = electrons.back();
 
-  if (!eleCore->ecalDrivenSeed()) {
-    delete eleCore;
+  if (!eleCore.ecalDrivenSeed()) {
+    electrons.pop_back();
```

Comment:
Why not checking first if ```gsfTrackRef.ecalDrivenSeed()```, and then ```emplace_back``` only in that case?

Reply:
Good idea, but it turns out the `GsfTrack` does not have this member function, so one would have to reproduce the logic from the GsfElectronCore producer I think [1]. I'm not really sure that we can find a quick and safe solution without code duplication in the scope of this PR, or do you have any idea?

[1] https://github.com/cms-sw/cmssw/blob/master/DataFormats/EgammaCandidates/src/GsfElectronCore.cc#L15